wire-schema/src/jvmMain/java/com/squareup/wire/schema/EmittingRules.kt

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.schema
+
+/**
+ * A set of rules that describes which types to generate.
+ *
+ * Despite the builder, instances of this class are not safe for concurrent use.
+ *
+ * ### Identifier Matching
+ *
+ * Identifiers in this set may be in the following forms:
+ *
+ *  * Package names, followed by `.*`, like `squareup.protos.person.*`. This matches types and
+ *    services defined in the package and its descendant packages.
+ *
+ *  * Fully qualified type and service names, like `squareup.protos.person.Person`.
+ *
+ * Identifiers should not contain member names.
+ *
+ * This set has *included identifiers* and *excluded identifiers*, with excludes taking precedence
+ * over includes. That is, if a type `Movie` is in both the includes and the excludes, it is not
+ * contained in the set.
+ *
+ * If the includes set is empty, that implies that all elements should be included. Use this to
+ * exclude unwanted types and members without also including everything else.
+ */
+class EmittingRules private constructor(builder: Builder) {
+  private val includes = builder.includes.toSet()
+  private val excludes = builder.excludes.toSet()
+  private val usedIncludes = mutableSetOf<String>()
+  private val usedExcludes = mutableSetOf<String>()
+
+  val isEmpty: Boolean
+    get() = includes.isEmpty() && excludes.isEmpty()
+
+  /** Returns true if `type` should be generated. */
+  fun includes(type: ProtoType) = includes(type.toString())
+
+  private fun includes(identifier: String): Boolean {
+    if (includes.isEmpty()) return !exclude(identifier)
```

Comment:
we could get rid of the exclude method if we were to set `includes` to `*` if empty on building.